log4j-api-java9/src/main/java/org/apache/logging/log4j/util/StackLocator.java

```diff
@@ -31,6 +33,7 @@
 
     private final static StackLocator INSTANCE = new StackLocator();
 
+    private final static ThreadLocal<FqcnCallerLocator> LOCATOR = ThreadLocal.withInitial(FqcnCallerLocator::new);
```

Comment:
Should we check if threadlocals are enabled before using this since it's not a standard type like StringBuilder?

Reply:
What do you mean - "check if threadlocals are enabled"? A ThreadLocal is a standard Java class...

Reply:
See `org.apache.logging.log4j.core.util.Constants.ENABLE_THREADLOCALS` and `org.apache.logging.log4j.util.Constants.ENABLE_THREADLOCALS`.

I think we disable most threadlocals in web-apps to avoid N tomcat webapps leaving thread-local state on each webserver thread. If memory serves, a ThreadLocal of a type from the standard library can be shared between all threads, even if they have loaded different log4j jars, but I'm having trouble finding the comment where I read that.

Reply:
Hi Carter, yes. Not sure which comment you were looking for, but in general ThreadLocals that hold a reference to an application class (e.g. a Log4j class) may cause memory leaks in application servers that keep their thread pools up when a web app is stopped and started. If the ThreadLocal only contains classes that are in the JDK library, such memory leaks will not occur.